### PR TITLE
[JIT] nvfuser CI fixes

### LINKED ISSUE #75029: 2022-03-31 disabled nvfuser tests
### 🐛 Describe the bug

- [x] #75122
- [x] test_int_tensor_input (M60, https://github.com/pytorch/pytorch/runs/5775045801?check_suite_focus=true)
- [x] (https://github.com/pytorch/pytorch/pull/75116 ) test_native_batch_norm_backward (M60, https://github.com/pytorch/pytorch/runs/5775045801?check_suite_focus=true)
- [x] (https://github.com/pytorch/pytorch/pull/75116 ) test_reduction_empty_axes (M60, https://github.com/pytorch/pytorch/runs/5775045801?check_suite_focus=true)
- [x] (opinfo) test_nvfuser_correctness_nn_functional_conv_transpose3d_cuda_float16 (M60 , https://github.com/pytorch/pytorch/runs/5775863924?check_suite_focus=true)
- [x] (opinfo) test_nvfuser_correctness_nn_functional_conv_transpose3d_cuda_float32 (M60 , https://github.com/pytorch/pytorch/runs/5775863924?check_suite_focus=true)
- [x] (https://github.com/pytorch/pytorch/pull/75116 ) test_context_manager_test (M60, https://github.com/pytorch/pytorch/runs/5775863870?check_suite_focus=true)
- [x] (https://github.com/pytorch/pytorch/pull/75116 ) test_register_fuser (M60, https://github.com/pytorch/pytorch/runs/5775863870?check_suite_focus=true)
- [x] (https://github.com/pytorch/pytorch/pull/75340) test_category_rule - fails sometimes, succeeds sometimes.

```
2022-03-31T17:11:07.8343008Z ======================================================================
2022-03-31T17:11:07.8345843Z ERROR [0.014s]: test_linear_1d_weight_mismatch_bias_dtype (__main__.TestCudaFuser)
2022-03-31T17:11:07.8347205Z ----------------------------------------------------------------------
2022-03-31T17:11:07.8347834Z Traceback (most recent call last):
2022-03-31T17:11:07.8348239Z   File "test_jit_cuda_fuser.py", line 3322, in test_linear_1d_weight_mismatch_bias_dtype
2022-03-31T17:11:07.8348595Z     jit_o = jitted(x, w, b)
2022-03-31T17:11:07.8349122Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 428, in prof_func_call
2022-03-31T17:11:07.8349661Z     return prof_callable(func_call, *args, **kwargs)
2022-03-31T17:11:07.8350336Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 425, in prof_callable
2022-03-31T17:11:07.8351373Z     return callable(*args, **kwargs)
2022-03-31T17:11:07.8355944Z RuntimeError: The following operation failed in the TorchScript interpreter.
2022-03-31T17:11:07.8356484Z Traceback of TorchScript (most recent call last):
2022-03-31T17:11:07.8357053Z   File "test_jit_cuda_fuser.py", line 3312, in t
2022-03-31T17:11:07.8357728Z         def t(x: torch.Tensor, w: torch.Tensor, b: torch.Tensor):
2022-03-31T17:11:07.8358665Z             o = torch.nn.functional.linear(x, w, b)
2022-03-31T17:11:07.8359542Z                 ~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
2022-03-31T17:11:07.8360079Z             return o.relu()
2022-03-31T17:11:07.8360774Z RuntimeError: mat2 must be a matrix, got 1-D tensor
2022-03-31T17:11:07.8361193Z 
2022-03-31T17:11:07.8361223Z 
2022-03-31T17:11:07.8361691Z ======================================================================
2022-03-31T17:11:07.8362326Z FAIL [0.006s]: test_int_tensor_input (__main__.TestCudaFuser)
2022-03-31T17:11:07.8363398Z ----------------------------------------------------------------------
2022-03-31T17:11:07.8364030Z Traceback (most recent call last):
2022-03-31T17:11:07.8364752Z   File "test_jit_cuda_fuser.py", line 4060, in test_int_tensor_input
2022-03-31T17:11:07.8365442Z     self._run_helper(t_jit, t, x)
2022-03-31T17:11:07.8366108Z   File "test_jit_cuda_fuser.py", line 154, in _run_helper
2022-03-31T17:11:07.8366768Z     self.assertGraphContainsExactly(jit_op.graph_for(*args), FUSION_GUARD, 1, consider_subgraphs=True)
2022-03-31T17:11:07.8367464Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/jit_utils.py", line 326, in assertGraphContainsExactly
2022-03-31T17:11:07.8367878Z     consider_subgraphs)
2022-03-31T17:11:07.8368360Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/jit_utils.py", line 320, in perform_assert
2022-03-31T17:11:07.8368781Z     graph, actual, kind, subgraph, expected))
2022-03-31T17:11:07.8369104Z AssertionError: graph(%x.1 : Tensor):
2022-03-31T17:11:07.8369413Z   %1 : bool = prim::Constant[value=0]()
2022-03-31T17:11:07.8369720Z   %2 : int[] = prim::Constant[value=[0]]()
2022-03-31T17:11:07.8370062Z   %3 : Tensor = aten::amax(%x.1, %2, %1) # test_jit_cuda_fuser.py:4057:23
2022-03-31T17:11:07.8370369Z   return (%3)
2022-03-31T17:11:07.8370503Z 
2022-03-31T17:11:07.8370734Z Error: graph contains 0 prim::CudaFusionGuard nodes (including subgraphs) but expected 1
2022-03-31T17:11:07.8370999Z 
2022-03-31T17:11:07.8371142Z ======================================================================
2022-03-31T17:11:07.8371486Z FAIL [0.012s]: test_native_batch_norm_backward (__main__.TestCudaFuser)
2022-03-31T17:11:07.8371946Z ----------------------------------------------------------------------
2022-03-31T17:11:07.8372293Z Traceback (most recent call last):
2022-03-31T17:11:07.8372652Z   File "test_jit_cuda_fuser.py", line 4195, in test_native_batch_norm_backward
2022-03-31T17:11:07.8373127Z     save_invstd, True, 1e-5, [True, True, True]), FUSION_GUARD)
2022-03-31T17:11:07.8373673Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/jit_utils.py", line 311, in assertGraphContains
2022-03-31T17:11:07.8374091Z     self.assertTrue(len(out_nodes) > 0)
2022-03-31T17:11:07.8374393Z AssertionError: False is not true
2022-03-31T17:11:07.8374576Z 
2022-03-31T17:11:07.8374694Z ======================================================================
2022-03-31T17:11:07.8375029Z FAIL [0.006s]: test_reduction_empty_axes (__main__.TestCudaFuser)
2022-03-31T17:11:07.8375493Z ----------------------------------------------------------------------
2022-03-31T17:11:07.8375838Z Traceback (most recent call last):
2022-03-31T17:11:07.8376169Z   File "test_jit_cuda_fuser.py", line 4047, in test_reduction_empty_axes
2022-03-31T17:11:07.8376506Z     self._run_helper(t_jit, t, x)
2022-03-31T17:11:07.8376834Z   File "test_jit_cuda_fuser.py", line 154, in _run_helper
2022-03-31T17:11:07.8377260Z     self.assertGraphContainsExactly(jit_op.graph_for(*args), FUSION_GUARD, 1, consider_subgraphs=True)
2022-03-31T17:11:07.8377929Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/jit_utils.py", line 326, in assertGraphContainsExactly
2022-03-31T17:11:07.8378338Z     consider_subgraphs)
2022-03-31T17:11:07.8378834Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/jit_utils.py", line 320, in perform_assert
2022-03-31T17:11:07.8379234Z     graph, actual, kind, subgraph, expected))
2022-03-31T17:11:07.8379550Z AssertionError: graph(%x.1 : Tensor):
2022-03-31T17:11:07.8379941Z   %1 : bool = prim::Constant[value=0]()
2022-03-31T17:11:07.8380231Z   %2 : NoneType = prim::Constant()
2022-03-31T17:11:07.8380560Z   %3 : int[] = prim::Constant[value=annotate(List[int], [])]()
2022-03-31T17:11:07.8380932Z   %4 : Tensor = aten::sum(%x.1, %3, %1, %2) # test_jit_cuda_fuser.py:4044:23
2022-03-31T17:11:07.8381272Z   return (%4)
2022-03-31T17:11:07.8381421Z 
2022-03-31T17:11:07.8381651Z Error: graph contains 0 prim::CudaFusionGuard nodes (including subgraphs) but expected 1
```

```
FAIL [0.063s]: test_nvfuser_correctness_nn_functional_conv_transpose3d_cuda_float16 (__main__.TestCudaFuserOpInfoCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1780, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1780, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 780, in test_wrapper
    return test(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1087, in wrapper
    fn(*args, **kwargs)
  File "/var/lib/jenkins/workspace/test/test_jit_cuda_fuser.py", line 4457, in test_nvfuser_correctness
    val = trace(*clone_inputs((sample.input, *sample.args)), **sample.kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/jit_metaprogramming_utils.py", line 368, in traced_fn
    self.assertExportImport(traced.graph, inputs_tensors)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_jit.py", line 148, in assertExportImport
    self.assertExportImportModule(m, inputs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_jit.py", line 154, in assertExportImportModule
    self.assertEqual(a, b, "Results of original model and "
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 2150, in assertEqual
    assert_equal(
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_comparison.py", line 1084, in assert_equal
    raise error_metas[0].to_error()
AssertionError: Results of original model and exported/imported version of model differed

======================================================================
FAIL [0.161s]: test_nvfuser_correctness_nn_functional_conv_transpose3d_cuda_float32 (__main__.TestCudaFuserOpInfoCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1780, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1780, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 780, in test_wrapper
    return test(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1087, in wrapper
    fn(*args, **kwargs)
  File "/var/lib/jenkins/workspace/test/test_jit_cuda_fuser.py", line 4455, in test_nvfuser_correctness
    trace(*clone_inputs((sample.input, *sample.args)), **sample.kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/jit_metaprogramming_utils.py", line 368, in traced_fn
    self.assertExportImport(traced.graph, inputs_tensors)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_jit.py", line 148, in assertExportImport
    self.assertExportImportModule(m, inputs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_jit.py", line 154, in assertExportImportModule
    self.assertEqual(a, b, "Results of original model and "
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 2150, in assertEqual
    assert_equal(
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_comparison.py", line 1084, in assert_equal
    raise error_metas[0].to_error()
AssertionError: Results of original model and exported/imported version of model differed

----------------------------------------------------------------------
Ran 4571 tests in 2673.458s

FAILED (failures=2, skipped=206, expected failures=6)
```

```
2022-03-31T18:09:00.4229799Z FAIL [0.513s]: test_context_manager_test (__main__.TestPassManagerCudaFuser)
2022-03-31T18:09:00.4230330Z ----------------------------------------------------------------------
2022-03-31T18:09:00.4230663Z Traceback (most recent call last):
2022-03-31T18:09:00.4231703Z   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1780, in wrapper
2022-03-31T18:09:00.4232092Z     method(*args, **kwargs)
2022-03-31T18:09:00.4232472Z   File "/var/lib/jenkins/workspace/test/test_jit_cuda_fuser.py", line 4422, in test_context_manager_test
2022-03-31T18:09:00.4233045Z     self.assertGraphContainsExactly(t_jit_3.graph_for(x, y), FUSION_GUARD, 0)
2022-03-31T18:09:00.4233690Z   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/jit_utils.py", line 339, in assertGraphContainsExactly
2022-03-31T18:09:00.4234158Z     perform_assert(graph, kind, len(out_nodes), num_kind_nodes,
2022-03-31T18:09:00.4234686Z   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/jit_utils.py", line 318, in perform_assert
2022-03-31T18:09:00.4235076Z     raise AssertionError(
2022-03-31T18:09:00.4235362Z AssertionError: graph(%x.1 : Tensor,
2022-03-31T18:09:00.4235635Z       %y.1 : Tensor):
2022-03-31T18:09:00.4236118Z   %2 : bool = prim::CudaFusionGuard[types=[Float(4, 8, strides=[8, 1], requires_grad=0, device=cuda:0), Float(4, 8, strides=[8, 1], requires_grad=0, device=cuda:0)]](%x.1, %y.1)
2022-03-31T18:09:00.4236553Z   %3 : Tensor = prim::If(%2)
2022-03-31T18:09:00.4236804Z     block0():
2022-03-31T18:09:00.4237102Z       %o.8 : Tensor = prim::CudaFusionGroup_0[cache_id=1650](%x.1, %y.1)
2022-03-31T18:09:00.4237453Z       -> (%o.8)
2022-03-31T18:09:00.4237684Z     block1():
2022-03-31T18:09:00.4237991Z       %5 : Function = prim::Constant[name="fallback_function", fallback=1]()
2022-03-31T18:09:00.4238425Z       %6 : (Float(4, 8, strides=[8, 1], requires_grad=0, device=cuda:0)) = prim::CallFunction(%5, %x.1, %y.1)
2022-03-31T18:09:00.4238858Z       %7 : Float(4, 8, strides=[8, 1], requires_grad=0, device=cuda:0) = prim::TupleUnpack(%6)
2022-03-31T18:09:00.4239209Z       -> (%7)
2022-03-31T18:09:00.4239421Z   return (%3)
2022-03-31T18:09:00.4239776Z with prim::CudaFusionGroup_0 = graph(%0 : Float(4, 8, strides=[8, 1], requires_grad=0, device=cuda:0),
2022-03-31T18:09:00.4240186Z       %1 : Float(4, 8, strides=[8, 1], requires_grad=0, device=cuda:0)):
2022-03-31T18:09:00.4240610Z   %2 : float = prim::Constant[value=4.]() # /var/lib/jenkins/workspace/test/test_jit_cuda_fuser.py:4417:20
2022-03-31T18:09:00.4240968Z   %3 : int = prim::Constant[value=1]()
2022-03-31T18:09:00.4241414Z   %o.1 : Float(4, 8, strides=[8, 1], requires_grad=0, device=cuda:0) = aten::add(%0, %1, %3) # /var/lib/jenkins/workspace/test/test_jit_cuda_fuser.py:4416:16
2022-03-31T18:09:00.4241974Z   %o.5 : Float(4, 8, strides=[8, 1], requires_grad=0, device=cuda:0) = aten::add(%o.1, %2, %3) # /var/lib/jenkins/workspace/test/test_jit_cuda_fuser.py:4417:16
2022-03-31T18:09:00.4242354Z   return (%o.5)
2022-03-31T18:09:00.4242492Z 
2022-03-31T18:09:00.4242722Z Error: graph contains 1 prim::CudaFusionGuard nodes (excluding subgraphs) but expected 0
2022-03-31T18:09:00.4242987Z 
2022-03-31T18:09:00.4243123Z ======================================================================
2022-03-31T18:09:00.4243475Z FAIL [0.002s]: test_register_fuser (__main__.TestPassManagerCudaFuser)
2022-03-31T18:09:00.4243946Z ----------------------------------------------------------------------
2022-03-31T18:09:00.4244300Z Traceback (most recent call last):
2022-03-31T18:09:00.4244810Z   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1780, in wrapper
2022-03-31T18:09:00.4245190Z     method(*args, **kwargs)
2022-03-31T18:09:00.4245540Z   File "/var/lib/jenkins/workspace/test/test_jit_cuda_fuser.py", line 4426, in test_register_fuser
2022-03-31T18:09:00.4245953Z     self.assertFalse(torch._C._jit_set_nvfuser_enabled(True))
2022-03-31T18:09:00.4246285Z AssertionError: True is not false
```

### Versions

Various CI machines, as mentioned above

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #75116

* test_native_batch_norm_backward
* test_reduction_empty_axes
* test_register_fuser
* test_category_rule

### CODE REVIEW

File: test/test_jit_cuda_fuser.py [hunk 1]
```diff
@@ -4388,11 +4387,19 @@ def fn(x, y):
             torch._C._jit_nvfuser_clear_comparison_callback()
 
 class TestPassManagerCudaFuser(JitTestCase):
+    def setUp(self):
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py:4390]: Looks like we have some other test that leaked the flag and accidentally left nvfuser as enabled, or is it some thread racing that other tests running in parallel are turning the nvfuser switch?

**Review by jjsjann123 — COMMENTED** [review]
LGTM, unfortunately the reduction issues are not repro'ing on my local Pascal card, maybe I got lucky with the heuristics....

**Review by jjsjann123 — APPROVED** [review]
forgot to stamp

File: test/test_jit_cuda_fuser.py (hunk 1 above)
Comment (davidberard98) [reply test/test_jit_cuda_fuser.py:4390]: another test leaked the flag (my fault, fixed it on line 159)

**Comment by davidberard98** [comment]
> LGTM, unfortunately the reduction issues are not repro'ing on my local Pascal card, maybe I got lucky with the heuristics....

@jjsjann123 I was assuming it was safe to skip since the other reduction tests are marked with the same, lmk if that's not accurate

**Comment by jjsjann123** [comment]
> > LGTM, unfortunately the reduction issues are not repro'ing on my local Pascal card, maybe I got lucky with the heuristics....
> 
> @jjsjann123 I was assuming it was safe to skip since the other reduction tests are marked with the same, lmk if that's not accurate

No it looks fine to me as well. I need to double check what went wrong with the repro on my local machine... This PR looks good to merge.

**Review by eellison — APPROVED** [review]
LGTM

**Comment by jjsjann123** [comment]
> LGTM, unfortunately the reduction issues are not repro'ing on my local Pascal card, maybe I got lucky with the heuristics....

Just for the record, I've been really dumb and somehow messed the device number on my machine. Unsurprisingly I've been running all my tests on a volta card instead of the pascal card.... No wonder why it does not repro....  :face_exhaling:

**Pull request closed without merging**